src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/CommandMessageBinaryEncoder.cs

```diff
@@ -288,25 +294,157 @@ private void WriteType1Section(BsonBinaryWriter writer, Type1CommandMessageSecti
             var payloadStartPosition = stream.Position;
             stream.WriteInt32(0); // size
             stream.WriteCString(section.Identifier);
+
             var batch = section.Documents;
-            var processedCount = batch.Count;
-            for (var i = 0; i < batch.Count; i++)
+            var maxDocumentSize = section.MaxDocumentSize ?? writer.Settings.MaxDocumentSize;
+            writer.PushSettings(s => ((BsonBinaryWriterSettings)s).MaxDocumentSize = maxDocumentSize);
+            writer.PushElementNameValidator(section.ElementNameValidator);
```

Comment:
When writing the payload we override the MaxDocumentSize and the ElementNameValidator.